opcua.cpp

```diff
@@ -1507,6 +1674,44 @@ SOPC_ClientHelper_GetEndpointsResult *OPCUA::GetEndPoints(const char *endPointUr
 	return endpoints;
 }
 
+
```

Comment:
I am concerned about the introduction of the static method _Node::getNodeAttr()_. It does the same job as the Node constructor but doesn't keep any of the returned data. Every call to _getNodeAttr_ or the _Node_ constructor results in a network call to the OPC UA server. We make one call per _NodeId_ (we should improve this in the future but that would be a new Jira). In the _OPCUA::subscribe_ method, there are 3 calls to the server for each _NodeId_: (i) Line 888 calls _checkNode_ which in turn calls _getNodeAttr_, (ii) Line 897 which constructs a _Node_ object, and (iii) Line 934 which calls _getNodeAttr_ directly. Please take advantage of the _OPCUA::m_nodes_ member which is a map from a string _NodeId_ to a _Node_ object. There is code in the _OPCUA::browse_ method that demonstrates how this should be used. See Lines 1875 to 1879. Check _m_nodes_ every time we need node information: if a _NodeId_ is found in the map, use the information in the _Node_ object. If not, get the node information by instantiating a new _Node_ object and then add it to the _m_nodes_ map.

Reply:
1. Sometimes, browse name is fetched from the server, then filtering check is done and if the node is to be excluded, a Node object is not created and hence no entry is added to m_nodes map. 
2. So if really a proper cache is required so that query for any node is done only once, it has to be different from m_nodes. Or we could have a partial cache in m_nodes.
3. Also this server query is done at startup only, and hence a few extra queries might be ok for now?
4. Also I have replaced one checkNode() call with checkFiltering() call, since getNodeAttr() was called a few lines above in subscribe().

Reply:
Regarding #3: No, it's not. When I was working on [FOGL-7849](https://scaledb.atlassian.net/browse/[FOGL-7849](https://scaledb.atlassian.net/browse/FOGL-7849)), I tested many configurations of 10,000 NodeIds or more after the _S2OPC Toolkit_ was fixed. Startup took between 5 and 7 minutes. I then discovered that the Toolkit's _SOPC_ClientHelper_Browse_ method called by _OPCUA::browse_ had the side-effect of retrieving the _BrowseName_ and _NodeClass_ of every _Variable_ it browsed, so I created a _Node_ object and put it into the _m_nodes_ map right away. Startup time went from 5-7 minutes to about 5 seconds. In my case, I had top-level folders in Kepware with all of my Variables inside it so almost all Variables were discovered by browsing. The original code had been using _OPCUA::browse_ to load the _variables_ string array with _NodeIds_ but then instantiating a _Node_ object later which requires a round trip to the server. My point is that anything that can be done to eliminate round trips will be beneficial. Since the startup time of the currently released version is quite good, I don't think we can justify reducing performance with a promise to fix it later.

Reply:
Regarding #2, the _m_nodes_ map is already a proper cache. It is referenced in _OPCUA::dataChange_ using the key (NodeId) which is passed to us by the S2OPC Toolkit. It would not matter if _m_nodes_ had extra Nodes in it but it doesn't need to. The _variables_ string vector is the definitive list the NodeIds of all Variables that need to be registered.

NodeIds are added to the variables array (using _variables.push_back()_) in only two places:

- In _OPCUA::subscribe_ at Line 915 of your current branch, and
- In _OPCUA::browse_ at Line 1886. 

All you need to do is add an entry to _m_nodes_ at the same time you add an entry to the variables vector. The _OPCUA::browse_ method already does this.

Method _OPCUA::subscribe_ at Line 915 does not. If Line 912 is changed to dynamically instantiating a Node object (i.e. _Node *node = new Node(...)_), then you could insert it into _m_nodes_ right after the _variables.push_back_. You could do even better by restructuring the _m_subscriptions_ loop (Lines 901 through 926). If you instantiate a Node object at the top of the loop, you can use the object to check for filtering status. If you need to keep it, you add it to the _m_nodes_ map right after _variables.push_back_. Otherwise, delete it. It would result in fewer round trips to server than right now.

I would make only a few changes to the _variables_  loop (Lines 945 to 995). Right at the top of the loop, rather than calling _Node:::getNodeAttr_, try to retrieve a Node object from the _m_nodes_ map. It should already be there. If it isn't, create one. The code that currently adds Node objects to the _m_nodes_ map (Lines 966 to 973) checks the map first so it's harmless.

By the way, I don't think checking filtering in the _variables_ loop would ever find anything (Lines 953 to 962). Filtering is already checked in the _m_subscriptions_ loop, either directly or within _OPCUA::browse_. I suppose it's harmless if there won't be a need a round-trip to server.

Reply:
Made changes in code to maintain and use cache in m_nodes. Existing comments seemed to suggest m_nodes contained the nodes that were subscribed to, hence my confusion. Removed getNodeAttr() and checkNode() functions.